### PR TITLE
Add TCP Fast Open support

### PR DESCRIPTION
**Pull request by wongsyrone**
Forward proxy has been removed since we cannot control
its socket options.

Signed-off-by: Syrone Wong <wong.syrone@gmail.com>

### CODE REVIEW

**Comment by wongsyrone** [comment]
It hasn't been tested on Windows 7; I don't have such devices.

**Comment by chenshaoju** [comment]
The Appveyor url in readme.md is broken ( https://ci.appveyor.com/project/wongsyrone/shadowsocks-windows-yqdou ) Unable test.

(Too lazy for compile 😅 )

**Comment by wongsyrone** [comment]
https://ci.appveyor.com/project/celeron533/shadowsocks-windows/build/4.0.5.24

**Comment by chenshaoju** [comment]
On Windows 7,Log print:
```
[2017-08-19 17:08:48] TFO: False
```

**Comment by wongsyrone** [comment]
It's expected. https://github.com/shadowsocks/shadowsocks-windows/issues/1045

**Comment by celeron533** [comment]
The requirement is conflick with pull request #1214 (Forward http proxy with Auth) :worried:
Also may have impact with #1298 (SIP002, SIP003)

**Comment by wongsyrone** [comment]
I talked to @madeye before, he agreed to remove Forward proxy.

**Comment by madeye** [comment]
Is it possible to implement forward proxy as a SIP003 plugin instead? If so, we can keep our core code base as clean as possible.

**Comment by madeye** [comment]
IMO, the priority list of pull requests is SIP003 > TFO > Forward proxy.

**Comment by wongsyrone** [comment]
Okay. This will remains a reference implementation for people who interested on both of them. I choose not to modify my PR.

**Comment by rwasef1830** [comment]
@madeye @wongsyrone  Why is this in conflict with my pull request ? I don't touch any of the core tcp / udp relay code (I implemented it in a far simpler way from a higher level).

I see no reason why we can't have tcp fast open at least for the supported cases (no forward proxy, no sip003 plugin).

**Comment by madeye** [comment]
Yes, if there is no conflict, I think we can try to merge this PR next.

@rwasef1830 Any interest to help merge this one?

**Comment by rwasef1830** [comment]
Yes, should I open a new pull request and cherry pick the commit to my fork ? I want to maintain credits to @wongsyrone

**Comment by madeye** [comment]
I think we'd better open a new pull request and rebase the latest changes then.

@wongsyrone What do you think?

**Comment by wongsyrone** [comment]
There are two must-have for TFO.

1. be able to control remote socket option
2. be able to receive handshaking from client thus we have data to send when connecting

Under forward proxy, I have no idea on fulfilling these two requirements.

Currently, I'm investigating c-ares, cannot find much time to rebase. If someone has interest, he can mention my ID in the commit message. As I said before, it serves as a reference implementation, everyone can use without notice.

**Comment by rwasef1830** [comment]
@wongsyrone ok, thanks. I will cherry pick and open a new pull request.

**Comment by celeron533** [comment]
Sorry I have not get sufficient time to review all the code of this pull request and do a full assessment of code impacts.

**Comment by wongsyrone** [comment]
Forgot to mention, the two must-have above is specific to the Socket.ConnectAsync() way. If you want to pass parameters to the plugin like: "param=tcp-fast-open;blah-blah" to notify other layers to enable TFO. That layer should use ConnectEx on Windows, otherwise, it won't work.

**Comment by rwasef1830** [comment]
@wongsyrone Yeah I noticed this. Currently, I am trying to make the changes as small as possible to implement this, but I really have the urge of rewriting that layer.. The ancient BeginXXX EndXXX style code is very messy and hard to follow.

**Comment by wongsyrone** [comment]
Do you want to use TAP? I already implemented it with SAEA for the whole networking layer.

**Comment by rwasef1830** [comment]
I want to change IProxy interface to expose TAP and wrap SAEA into TAP interface, it is much cleaner. Even better if I can make it expose Stream interface, so unit testing becomes possible.

**Comment by wongsyrone** [comment]
In fact, there is no need to convert all code into TAP. If you use SAEA, there is no need to keep connect timer, just let the socket layer handle it.
